File: ruby/collatz-conjecture/collatz_conjecture.rb
# frozen_string_literal: true
class CollatzConjecture

  def self.steps(limit)

    throw ArgumentError if limit < 1

    cnt = 0

    while limit > 1
      limit = limit.even? ? limit / 2 : 3 * limit + 1
      cnt += 1
    end

    cnt

  end
end

File: fortran/collatz-conjecture/collatz_conjecture.f90
module collatz_conjecture
  implicit none
contains
  integer function steps(limit)
  integer, intent(in) :: limit
  integer :: n
  n = limit
  steps = -1
  if (n.lt.1) return
  steps = 0
  do while (n.ne.1)
    if (mod(n,2).eq.0) then
      n = n / 2
    else
      n = n * 3 + 1
    end if
    steps = steps + 1
  end do
  end function
end module

